### PR TITLE
CSharp Interop

### PR DESCRIPTION
**Pull request by jannesiera**
- Two new projects (`Dash.NET.CSharp` and `Dash.NET.CSharp.Giraffe`) have been created as an adapter to wrap the original F# implementation and expose an API that is easier to consume from C#. C# users should not be exposed to the original `Dash.NET` and `Dash.NET.Giraffe` namespaces but only have access to the `Dash.NET.CSharp` and `Dash.NET.CSharp.Giraffe` namespaces.
 - Two new example projects (`Dash.NET.Giraffe.CSharp.Example` and `Dash.NET.Giraffe.CSharp.Documentation.Examples`) have been created to demonstrate how the API is been used from C#. The second of which holds all examples used by the documentation. These will be expanded upon as we continue writing the C# documentation.

### CODE REVIEW

**Review by ademar — COMMENTED** [review]
Hi Janne,  This looks good to me,  very clean and professional.

My only comment is about the overall approach to the problem and probably outside the scope of this PR. I think that if we want to facilitate the consumption of the library from C# we should code the original components with that in mind (avoid using options, DUs and other F# idiomatics) instead of having to maintain a C# shim for each component which would be extra work to maintain. I understand that now is probably too late for that but something to consider for the future.

Other than that I have no complaints about your work and if everyone agrees this can be merged.

**Comment by kMutagene** [comment]
As we have discussed most of this previously, i like all of this.

I have some reservations about the approach of adding an overload for callbacks with increasing parameter numbers to the F# side though (also is this necessary in C# or just the 'C# way' to interop with callbacks?). I understand where this is coming from, but i think it is necessary accept at some point that not everything in this lib can be 100% type safe in a descriptive way while still being comfortable to use due to the requirements of the underlying infrastructure on which we have no influence. Dash is inherently dynamic, and all other language versions are dynamically typed. I have tried to stay typesafe in my original implementation wherever possible, but i think callbacks need to be as they are now to stay true to the original (well, they are typesafe, but the generic type for callbacks make it possible to create callbacks that wont work).

If there is a real alternative i am very open to it, but it is certainly not feasible to add 200 overloads just to have that one user that actually needs 201 inputs - it may be necessary in C# though, i do not know enough about it. I think this situation can be mitigated by adding good documentation, especially design notes about the library and usage of callbacks. We could offer both approaches though. 

> I think that if we want to facilitate the consumption of the library from C# we should code the original components with that in mind (avoid using options, DUs and other F# idiomatics) instead of having to maintain a C# shim for each component which would be extra work to maintain

Strongly disagree with this one. This is a F# library with C# support, not the other way around. I would be sad to see that switching. As components should be auto generated from react metadata, adding that shim generation should not be too much of an overhead. It is only now that the generator is not finished that this had to be done by hand.

**Review by kMutagene — APPROVED** [review]
LGTM overall, see my other comments in the PR thread. It would be really nice if you could add a test lib for the C# version that mirrors the endpoint tests. If this escapes the scope of this PR though i would say this is ready to merge (with the added note that we REALLY need some tests ;D)

**Review by jackparmer — APPROVED** [review]

**Comment by jackparmer** [comment]
Hey @jonsequitur @TIHan @colombod tagging you in case you're interested in reviewing 👀 
This is a pretty major PR that will bring Dash to C# (Dash is [currently only available in F#](https://dash.plotly.com/fsharp))
No pressure of course - just in case there are implications for .NET interactive or other projects where you would like to see Dash C#
Cheers! 🍻

File: Dash.NET.Giraffe.CSharp.Example/Program.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    Attr.children(
+                        Graph.graph(id: "my-graph-id"),
+                        Html.p(Attr.children("Petal Width:")),
+                        RangeSlider.rangeSlider(
+                            id: "range-slider",
+                            RangeSlider.Attr.min(0),
+                            RangeSlider.Attr.max(2.5),
+                            RangeSlider.Attr.step(0.1),
+                            RangeSlider.Attr.marks(
+                                new Dictionary<double, RangeSlider.Mark>()
```
Comment (colombod) [inline Dash.NET.Giraffe.CSharp.Example/Program.cs:87]: I like this a lot!

File: Dash.NET.Giraffe.CSharp.Tests/Layout_CoreComponents.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+namespace Documentation.Examples
+{
+    class Layout_CoreComponents
+    {
+        public static void RunExample()
+        {
+            var layout =
+                Html.div(
+                    Attr.children(
```
Comment (colombod) [inline Dash.NET.Giraffe.CSharp.Tests/Layout_CoreComponents.cs:17]: the attr to set childres seems a little bit too much, elements that have childred could be constructed using the params approach in csharp, then use the Attr.childred internally?

File: Dash.NET.Giraffe.CSharp.Tests/Layout_ReusableComponents.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public static void RunExample()
+        {
+            var csv = new HttpClient().GetStringAsync("https://gist.githubusercontent.com/chriddyp/c78bf172206ce24f77d6363a2d754b59/raw/c353e8ef842413cae56ae3920b8fd78468aa4cb2/usa-agricultural-exports-2011.csv").Result;
+            var rows = csv.Split("\n").SkipLast(1).Select(x => x.Split(","));
+            var headers = rows.First();
+            var data = rows.Skip(1);
+
+            var layout =
+                Html.div(
+                    Attr.children(
```
Comment (colombod) [inline Dash.NET.Giraffe.CSharp.Tests/Layout_ReusableComponents.cs:21]: is there any less verbose approach we can put in place, pretty sure that setting child elements to dom elements is the default thing every user will do

File: dev/Program.fs [hunk 4]
```diff
@@ -40,7 +40,7 @@ let dslLayout =
                 Dropdown.Attr.multi true
             ]
             Html.br []
-            Html.label [ Attr.children (Html.text "Selected values :") ]
```
Comment (colombod) [inline dev/Program.fs:43]: childred in html dom is regarded as property not as attributes, 
I found this, not sure if it makes sense for the api surface
image

**Review by colombod — APPROVED** [review]

**Pull request merged**